doc/paper/paper.tex

```diff
@@ -0,0 +1,1011 @@
+		filename, offset, and length.
+
+		They byte identifier of the WAL record takes one of three types, indicating
+		how the data should be parsed: \textbf{Insert} contains a serialized partial
+		tree covering one leaf node.  \textbf{MergeRequest} contains a list of insert
+		record addresses that should be merged as a batch into storage, as well as a
+		UUID-format batch ID.  \textbf{MergeComplete} records the completion of a
+		merge operation into storage by batch ID.
+
+		On startup, the WAL manager scans all outstanding WAL files in the WAL
+		directory, replaying their records through its accounting until it arrives at a
+		final state representing unmerged work in the WAL. Thereafter, writes to the
+		WAL are accounted for in memory and this state is used for scheduling of
+		flushes to storage.
+
+		Garbage collection of the WAL is handled with a process that polls the
+		in-memory state of the WAL manager to determine if any files on disk are no
+		longer referenced. Any such files are deleted.
+
+		\subsection{Rootmap}
+		The rootmap fronts a database of root locations used by the tree manager. Every
+		write that is finalized to storage records a write to the rootmap, and every
+		read from dp3 (when operating in server mode) does a read from the rootmap.
+		While the rootmap is technically rebuildable from storage operations, it is
+		essential for fast data access and effectively a central point of failure.
+
+		The rootmap interface is fairly straightforward and should be portable to a
+		range of database solutions:
+
+		\textbf{Get(database string, producer string, topic string, version uint64)}
+		gets a root for in a database a table at a version.
+
+		\textbf{GetLatestByTopic(database string, producer string, topics
+		map[string]uint64)} gets the latest roots in a database for a set of tables.
+		The passed map of topics indicates minimum tree versions for each topic.
+
+		\textbf{GetHistorical(database string, producer string, topic string)} gets
+		all root versions in the database for a single table.
+
+		\textbf{Put(database string, producer string, topic string, version uint64,
+		nodeID nodeID)} inserts a new root into the rootmap.
+
+		One dp3 instance can contain multiple logical \q{databases}, allowing users to
+		segregate simulation data from real-world data or maintain other such
+		divisions. All operations against the rootmap are therefore scoped with a
+		logical database identifier.
+
+		Today, dp3 uses SQLite for the rootmap.
+
+		\subsection{Versionstore}
+		For a single tree, writes must be versioned in monotonically increasing order
+		to support range deletion as well as querying for new data since a prior read.
+
+		The version store is backed by a persistent counter, from which reservations
+		of one billion versions are taken at a time. On startup, a dp3 instance
+		acquires a lock on the counter and increments it by a billion, thus reserving
```

Comment:
When I was doing my thesis i was told not to use words for numbers over 10, consider `1*10^9` or 1,000,000,000. Also historical note - [very old british english speakers](https://web.archive.org/web/20170112163426/https://en.oxforddictionaries.com/explore/how-many-is-a-billion) sometimes complain that billion used to mean 1e12.